cpp/ql/lib/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll

```diff
@@ -376,13 +380,13 @@ module TaintedWithPath {
    */
   class TaintTrackingConfiguration extends TSingleton {
     /** Override this to specify which elements are sources in this configuration. */
-    predicate isSource(Expr source) { exists(getNodeForSource(source)) }
+    predicate isSource(Element source) { exists(getNodeForSource(source)) }
```

Comment:
This change (and perhaps a few related ones) looks like it will break existing user code.  Is there a way we can keep existing code that overrides `isSource(Expr source)` working?

Reply:
Yeah, that's also what I'm worried about. I haven't yet found a good hack around this that would allow us to keep the `isSource(Expr)` working.

We introduced this `TaintTrackingConfiguration` mechanism in https://github.com/github/codeql/pull/3189 to get path explanations to the old queries that still used the `security.TaintTracking` library (which was changed to map to `DefaultTaintTracking`). As `@jbj` mentioned in [here](https://github.com/github/codeql/pull/3189#issuecomment-607645777) new queries really shouldn't be importing `security.TaintTracking`. The QLDoc on the `security.TaintTracking` file even says that users should prefer to use `semmle.code.cpp.dataflow.TaintTracking` when designing new queries. And luckily, since actual `DataFlow::Configurations` (which is the one you write if you're using the `semmle.code.cpp.dataflow.TaintTracking` library) doesn't have this `Expr` vs. `Element` problem since that's working purely with `DataFlow::Node`s instead.

With all that said, I'm still in the process a couple of small bugs. Once I've fixed those, I'll go back and think about a solution that has better backwards compatibility.